lib/errors.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+/**
+ * @module Errors
+ */
+/**
+ * Factory functions to create throwable error objects
+ */
+
+/**
+ * @summary
+ * File/s of test could not be found.
+ *
+ * @public
+ */
+function NoFilesMatchPatternError(message, pattern) {
+  var errorWithCode = new Error(message);
+  errorWithCode.code = 'ERR_MOCHA_NO_FILES_MATCH_PATTERN';
+  errorWithCode.pattern = pattern;
+  return errorWithCode;
+}
+
+/**
+ * @summary
+ * Callback for a running suite was not found.
+ *
+ * @public
+ */
+function MissingCallbackError(message) {
+  var errorWithCode = new Error(message);
+  errorWithCode.code = 'ERR_MOCHA_MISSING_CALLBACK';
+  return errorWithCode;
+}
+/**
+ * @summary
+ * Reporter specified in options not found.
+ *
+ * @public
+ */
+function InvalidReporterError(message) {
+  var errorWithCode = new Error(message);
+  errorWithCode.code = 'ERR_MOCHA_INVALID_REPORTER';
+  return errorWithCode;
+}
+
+/**
+ * @summary
+ * Interface specified in options not found.
+ *
+ * @public
+ */
+function InvalidInterfaceError(message) {
+  var errorWithCode = new Error(message);
+  errorWithCode.code = 'ERR_MOCHA_INVALID_INTERFACE';
+  return errorWithCode;
+}
+
+/**
+ * @summary
+ * Type of output specified was not supported.
+ *
+ * @public
+ */
+function NotSupportedError(message) {
+  var errorWithCode = new Error(message);
+  errorWithCode.code = 'ERR_MOCHA_NOT_SUPPORTED';
+  return errorWithCode;
+}
+
+/**
+ * @summary
+ * An argument was missing.
+ *
+ * @public
+ */
+function MissingArgumentError(message, argument) {
```

Comment:
All calls to `MissingArgumentError` should be converted to [`TypeError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/More_arguments_needed)

Reply:
I disagree that we need to consume `TypeError`, `ReferenceError`, etc., as a feature of a `code` prop is that Mocha nor its consumers needn't worry about what the constructor was.  They would have to check the `code` anyway, and are unlikely to care whether the constructor was a specialized built-in `Error` subclass.

Reply:
There is _no_ point in us creating a specialized error class when the existing `TypeError` with `ERR_INVALID_ARG_TYPE'` 'is _exactly_ what should be thrown for missing arguments.